src/primitives/compact_action.rs

```diff
@@ -0,0 +1,181 @@
+//! Defines actions for Orchard shielded outputs and compact action for light clients.
+
+// Review hint: this file is largely derived from src/note_encryption.rs
+
+use core::fmt;
+
+use zcash_note_encryption::{note_bytes::NoteBytes, EphemeralKeyBytes, ShieldedOutput};
+
+use crate::{
+    action::Action,
+    note::{ExtractedNoteCommitment, Nullifier, Rho},
+};
+
+use super::{orchard_domain::OrchardDomain, orchard_primitives::OrchardPrimitives};
+
+impl<A, P: OrchardPrimitives> ShieldedOutput<OrchardDomain<P>> for Action<A, P> {
+    fn ephemeral_key(&self) -> EphemeralKeyBytes {
+        EphemeralKeyBytes(self.encrypted_note().epk_bytes)
+    }
+
+    fn cmstar(&self) -> &ExtractedNoteCommitment {
+        self.cmx()
+    }
+
+    fn cmstar_bytes(&self) -> [u8; 32] {
+        self.cmx().to_bytes()
+    }
+
+    fn enc_ciphertext(&self) -> Option<&P::NoteCiphertextBytes> {
+        Some(&self.encrypted_note().enc_ciphertext)
+    }
+
+    fn enc_ciphertext_compact(&self) -> P::CompactNoteCiphertextBytes {
+        P::CompactNoteCiphertextBytes::from_slice(
+            &self.encrypted_note().enc_ciphertext.as_ref()[..P::COMPACT_NOTE_SIZE],
+        )
+        .unwrap()
+    }
+}
+
```

Comment:
Bug: missing trait impl that was not moved from `src/note_encryption.rs`:
```suggestion

impl ShieldedOutput<OrchardDomain, ENC_CIPHERTEXT_SIZE> for crate::pczt::Action {
    fn ephemeral_key(&self) -> EphemeralKeyBytes {
        EphemeralKeyBytes(self.output().encrypted_note().epk_bytes)
    }

    fn cmstar_bytes(&self) -> [u8; 32] {
        self.output().cmx().to_bytes()
    }

    fn enc_ciphertext(&self) -> &[u8; ENC_CIPHERTEXT_SIZE] {
        &self.output().encrypted_note().enc_ciphertext
    }
}

```
(fix it as necessary to be correct for the updated `ShieldedOutput` trait).